src/Database/Expression/WindowExpression.php

```diff
@@ -23,21 +23,154 @@
 /**
  * This represents a SQL window expression used by aggregate and window functions.
  */
-class WindowExpression implements ExpressionInterface
+class WindowExpression implements ExpressionInterface, WindowInterface
 {
+    /**
+     * @var \Cake\Database\ExpressionInterface[]
+     */
+    protected $partitions = [];
+
+    /**
+     * @var \Cake\Database\Expression\OrderByExpression|null
+     */
+    protected $order;
+
+    /**
+     * @inheritDoc
+     */
+    public function partition($partitions)
+    {
+        if (!$partitions) {
+            return $this;
+        }
+
+        if (!is_array($partitions)) {
+            $partitions = [$partitions];
+        }
+
+        foreach ($partitions as &$partition) {
+            if (is_string($partition)) {
+                $partition = new IdentifierExpression($partition);
+            }
+        }
+
+        /** @var \Cake\Database\ExpressionInterface[] $partitions */
+        $this->partitions = array_merge($this->partitions, $partitions);
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function order($fields)
+    {
+        if (!$fields) {
+            return $this;
+        }
+
+        if ($this->order === null) {
+            $this->order = new OrderByExpression();
+        }
+
+        $this->order->add($fields);
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function range(?int $start, ?int $end = 0)
+    {
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function rows(?int $start, ?int $end = 0)
+    {
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function groups(?int $start, ?int $end = 0)
+    {
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function excludeCurrent()
+    {
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function excludeGroup()
+    {
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function excludeTies()
+    {
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function excludeNoOthers()
+    {
+        return $this;
+    }
+
     /**
      * @inheritDoc
      */
     public function sql(ValueBinder $generator): string
     {
-        return 'OVER ()';
+        $partitionSql = '';
+        if ($this->partitions) {
+            $partitions = array_map(function ($partition) use ($generator) {
+                return $partition->sql($generator);
+            }, $this->partitions);
```

Comment:
Using a foreach would yield better performance, but it likely doesn't matter here.

Reply:
Sure, can do.